### DISCUSSION TITLE
Why is a local cache not recommended for production use?

### DISCUSSION THREAD

**Discussion by MT927** [keycloak/keycloak#30943] [Q&A] [answered]
Hello all, I have what seems to be an unusual configuration that is not described in the documentation, and was wondering if anyone has advice.

I am running multiple Keycloak instances on the same network, and they can reach each other via multicast. However, they are completely separate entities doing different work and **should not** form into an Infinispan cluster. I would like to totally disable the distributed cache feature.

I found in the [documentation](https://www.keycloak.org/server/caching) that "Keycloak uses only local caches and distributed caches are completely disabled by implicitly setting the --cache=local option. The local cache mode is intended only for development and testing purposes." The --cache=local option solves my problems when I try it, no servers attempt to form a cluster.

I also looked through the source and found quarkus\runtime\src\main\resources\cache-local.xml. This file has no distributed caches, all caches are local. When I use this with --cache=ispn --cache-config-file=cache-local.xml, the problem is also resolved. The log output is exactly the same as when I ran with --cache=local.

My questions are:
1. Why is --cache=local discouraged for production use? The documentation describes that several types of data are cached locally, indeed the default cache-ispn.xml has some local caches. Is caching everything locally bad in my case?
2. Is --cache=ispn --cache-config-file=cache-local.xml equivalent to --cache=local? If so, is it also discouraged for production use or does it function differently?

I did read through the source code and previous discussions to try to find answers for these questions, but did not find anything conclusive.

Thank you!

July 10th 2024: I also found this, someone who has a lot of Keycloak experience suggests --cache=local is acceptable in production use for a single-node Keycloak. Is this accurate? https://keycloak.discourse.group/t/config-global-state-for-standalone-keycloak-22-0-5/24022

**Comment by shawkins** [comment, marked as answer]
> Is this accurate?

Yes, but single-node Keycloaks is typically not recommended for production due to availability and potentially session loss. There is a preview feature in KC 25 for session persistence that can help prevent session loss.

**Reply by MT927** [reply]
Thanks for this information. The session loss concern, I imagine, is because if there is only one Keycloak node and it goes down then sessions are lost. Is there more to that concern? I am not using Keycloak at scale, only a small number of users and sessions, so running two instances always seemed like unnecessary complexity. 

I see the new feature at --features=persistent-user-sessions, soon I will upgrade to Keycloak 25.x.x and test this out! Thanks.

**Reply by shawkins** [reply]
> Is there more to that concern?

Not really. If it's an acceptable trade-off for you to run a single instance, then you can do that.

**Comment by metronom72** [comment]
Does it possible to utilise shared fs? to apprehend local caching across multiple nodes?

**Reply by rausub** [reply]
it uses embedded infinispan from my undderstanding
